node/src/components/storage.rs

```diff
@@ -1212,6 +1234,49 @@ fn check_block_metadata_db(
             ),
         }
     }
-    info!("Check for block_metadata_db complete");
+
+    drop(cursor);
+    txn.commit()?;
+
+    info!("block metadata database initialized");
+    Ok(())
+}
+
+/// Purges stale entries from the deploy metadata database.
+fn initialize_deploy_metadata_db(
+    env: &Environment,
+    deploy_metadata_db: &Database,
+    deleted_block_hashes: &HashSet<BlockHash>,
+) -> Result<(), LmdbExtError> {
+    info!("initializing deploy metadata database");
+    let mut txn = env.begin_rw_txn()?;
+    let mut cursor = txn.open_rw_cursor(*deploy_metadata_db)?;
+
+    for (raw_key, raw_val) in cursor.iter() {
+        let mut deploy_metadata: DeployMetadata = lmdb_ext::deserialize(raw_val)?;
+
+        let mut removed = false;
+        for block_hash in deleted_block_hashes {
+            if deploy_metadata
+                .execution_results
+                .remove(block_hash)
+                .is_some()
```

Comment:
Wouldn't it be better, in terms of performance, to check it the other way around? Instead of checking every block hash we removed for existence in the `deploy_metadata.execution_results` (that test should be `true` only for one block hash in that collection, so we're making a lot of wasteful iterations), we check if `delete_block_hashes` exists in `delete_block_hashes`.

Reply:
I think you're right in that we could assume `execution_results` will only ever have 0 or 1 entry, but I'd rather leave this as is for now just to be safe.  I'm hoping we soon get to fix my initial mistake in making the `execution_resuilts` a `HashMap`, at which time, this code will automatically become optimal.

Reply:
It won't become optimal anyway. Here, for every element (deploy) under the cursor (`m` elements) we iterate over all delete blocks again (`n` elements). So we're making `m x n` iterations. We need to change the second check to
```rust
for b in deploy_metadata.execution_results {
  if deleted_block_hashes.contains(b) {
    deploy_metadata.execution_results.remove(b);
  }
}
```
then it makes far fewer iterations and when we change the structure of `execution_results` to only ever contain single block, then that code becomes optimal.

Reply:
So, leaving the future changes out of the discussion, I think for now we need to keep `m x n` iterations to be safe.

Reply:
Not really, my variant would still be vastly better b/c it would iterate `m` times (for every deploy) and then only as many times as there are elements in the `deploy_metadata.execution_results` (which should be max 1 anyway).